README.md
[![Documentation Status](https://readthedocs.org/projects/pycloudwave/badge/?version=latest)](https://pycloudwave.readthedocs.io/)
[![image](https://coveralls.io/repos/PyMySQL/PyMySQL/badge.svg?branch=main&service=github)](https://coveralls.io/github/PyMySQL/PyMySQL?branch=main)

# PyMySQL

This package contains a pure-Python MySQL client library, based on [PEP
249](https://www.python.org/dev/peps/pep-0249/).

## Requirements

- Python -- one of the following:
  - [CPython](https://www.python.org/) : 3.7 and newer
  - [PyPy](https://pypy.org/) : Latest 3.x version
- MySQL Server -- one of the following:
  - [MySQL](https://www.mysql.com/) \>= 5.7
  - [MariaDB](https://mariadb.org/) \>= 10.3

## Installation

Package is uploaded on [PyPI](https://pypi.org/project/PyMySQL).

You can install it with pip:

    $ python3 -m pip install PyMySQL

To use "sha256_password" or "caching_sha2_password" for authenticate,
you need to install additional dependency:

    $ python3 -m pip install PyMySQL[rsa]

To use MariaDB's "ed25519" authentication method, you need to install
additional dependency:

    $ python3 -m pip install PyMySQL[ed25519]

## Documentation

Documentation is available online: <https://pycloudwave.readthedocs.io/>

For support, please refer to the
[StackOverflow](https://stackoverflow.com/questions/tagged/pycloudwave).

## Example

The following examples make use of a simple table

``` sql
CREATE TABLE `users` (
    `id` int(11) NOT NULL AUTO_INCREMENT,
    `email` varchar(255) COLLATE utf8_bin NOT NULL,
    `password` varchar(255) COLLATE utf8_bin NOT NULL,
    PRIMARY KEY (`id`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin
AUTO_INCREMENT=1 ;
```

``` python
import pycloudwave.cursors

# Connect to the database
connection = pycloudwave.connect(host='localhost',
                             user='user',
                             password='passwd',
                             database='db',
                             cursorclass=pycloudwave.cursors.DictCursor)

with connection:
    with connection.cursor() as cursor:
        # Create a new record
        sql = "INSERT INTO `users` (`email`, `password`) VALUES (%s, %s)"
        cursor.execute(sql, ('webmaster@python.org', 'very-secret'))

    # connection is not autocommit by default. So you must commit to save
    # your changes.
    connection.commit()

    with connection.cursor() as cursor:
        # Read a single record
        sql = "SELECT `id`, `password` FROM `users` WHERE `email`=%s"
        cursor.execute(sql, ('webmaster@python.org',))
        result = cursor.fetchone()
        print(result)
```

This example will print:

``` python
{'password': 'very-secret', 'id': 1}
```

## Resources

- DB-API 2.0: <https://www.python.org/dev/peps/pep-0249/>
- MySQL Reference Manuals: <https://dev.mysql.com/doc/>
- MySQL client/server protocol:
  <https://dev.mysql.com/doc/internals/en/client-server-protocol.html>
- "Connector" channel in MySQL Community Slack:
  <https://lefred.be/mysql-community-on-slack/>
- PyMySQL mailing list:
  <https://groups.google.com/forum/#!forum/pycloudwave-users>

## License

PyMySQL is released under the MIT License. See LICENSE for more
information.


CHANGELOG.md
# Changes

## v1.0.3

Release date: TBD

* Dropped support of end of life MySQL version 5.6
* Dropped support of end of life MariaDB versions below 10.3
* Dropped support of end of life Python version 3.6


## v1.0.2

Release date: 2021-01-09

* Fix `user`, `password`, `host`, `database` are still positional arguments.
  All arguments of `connect()` are now keyword-only. (#941)


## v1.0.1

Release date: 2021-01-08

* Stop emitting DeprecationWarning for use of ``db`` and ``passwd``.
  Note that they are still deprecated. (#939)
* Add ``python_requires=">=3.6"`` to setup.py. (#936)


## v1.0.0

Release date: 2021-01-07

Backward incompatible changes:

* Python 2.7 and 3.5 are not supported.
* ``connect()`` uses keyword-only arguments. User must use keyword argument.
* ``connect()`` kwargs ``db`` and ``passwd`` are now deprecated; Use ``database`` and ``password`` instead.
* old_password authentication method (used by MySQL older than 4.1) is not supported.
* MySQL 5.5 and MariaDB 5.5 are not officially supported, although it may still works.
* Removed ``escape_dict``, ``escape_sequence``, and ``escape_string`` from ``pycloudwave``
  module. They are still in ``pycloudwave.converters``.

Other changes:

* Connection supports context manager API. ``__exit__`` closes the connection. (#886)
* Add MySQL Connector/Python compatible TLS options (#903)
* Major code cleanup; PyMySQL now uses black and flake8.


## v0.10.1

Release date: 2020-09-10

* Fix missing import of ProgrammingError. (#878)
* Fix auth switch request handling. (#890)


## v0.10.0

Release date: 2020-07-18

This version is the last version supporting Python 2.7.

* MariaDB ed25519 auth is supported.
* Python 3.4 support is dropped.
* Context manager interface is removed from `Connection`. It will be added
  with different meaning.
* MySQL warnings are not shown by default because many user report issue to
  PyMySQL issue tracker when they see warning. You need to call "SHOW WARNINGS"
  explicitly when you want to see warnings.
* Formatting of float object is changed from "3.14" to "3.14e0".
* Use cp1252 codec for latin1 charset.
* Fix decimal literal.
* TRUNCATED_WRONG_VALUE_FOR_FIELD, and ILLEGAL_VALUE_FOR_TYPE are now
  DataError instead of InternalError.


## 0.9.3

Release date: 2018-12-18

* cryptography dependency is optional now.
* Fix old_password (used before MySQL 4.1) support.
* Deprecate old_password.
* Stop sending ``sys.argv[0]`` for connection attribute "program_name".
* Close connection when unknown error is happened.
* Deprecate context manager API of Connection object.

## 0.9.2

Release date: 2018-07-04

* Disabled unintentinally enabled debug log
* Removed unintentionally installed tests


## 0.9.1

Release date: 2018-07-03

* Fixed caching_sha2_password and sha256_password raise TypeError on PY2
  (#700, #702)


## 0.9.0

Release date: 2018-06-27

* Change default charset from latin1 to utf8mb4.  (because MySQL 8 changed) (#692)
* Support sha256_password and caching_sha2_password auth method (#682)
* Add cryptography dependency, because it's needed for new auth methods.
* Remove deprecated `no_delay` option (#694)
* Support connection attributes (#679)
* Map LOCK_DEADLOCK to OperationalError (#693)

## 0.8.1

Release date: 2018-05-07

* Reduce `cursor.callproc()` roundtrip time. (#636)

* Fixed `cursor.query()` is hunged after multi statement failed. (#647)

* WRONG_DB_NAME and WRONG_COLUMN_NAME is ProgrammingError for now. (#629)

* Many test suite improvements, especially adding MySQL 8.0 and using Docker.
  Thanks to Daniel Black.

* Dropped support for old Python and MySQL which is not tested long time.


## 0.8

Release date: 2017-12-20

* **BACKWARD INCOMPATIBLE** ``binary_prefix`` option is added and off
  by default because of compatibility with mysqlclient.
  When you need PyMySQL 0.7 behavior, you have to pass ``binary_prefix=True``.
  (#549)

* **BACKWARD INCOMPATIBLE** ``MULTI_STATEMENTS`` client flag is no longer
  set by default, while it was on PyMySQL 0.7.  You need to pass
  ``client_flag=CLIENT.MULTI_STATEMENTS`` when you connect to explicitly
  enable multi-statement mode. (#590)

* Fixed AuthSwitch packet handling.

* Raise OperationalError for MariaDB's constraint error. (#607)

* executemany() accepts query without space between ``VALUES`` and ``(``.  (#597)

* Support config file containing option without value. (#588)

* Fixed Connection.ping() returned unintended value.


## 0.7.11

Release date: 2017-04-06

* Fixed Connection.close() failed when failed to send COM_CLOSE packet.
* Cursor.executemany() accepts query ends with semicolon.
* ssl parameters can be read from my.cnf.


## 0.7.10

Release date: 2017-02-14

* **SECURITY FIX**: Raise RuntimeError when received LOAD_LOCAL packet while
  ``loacal_infile=False``.  (Thanks to Bryan Helmig)

* Raise SERVER_LOST error for MariaDB's shutdown packet (#540)

* Change default connect_timeout to 10.

* Add bind_address option (#529)


## 0.7.9

Release date: 2016-09-03

* Fix PyMySQL stop reading rows when first column is empty string (#513)
  Reverts DEPRECATE_EOF introduced in 0.7.7.

## 0.7.8

Release date: 2016-09-01

* Revert error message change in 0.7.7.
  (SQLAlchemy parses error message, #507)

## 0.7.7

Release date: 2016-08-30

* Add new unicode collation (#498)
* Fix conv option is not used for encoding objects.
* Experimental support for DEPRECATE_EOF protocol.

## 0.7.6

Release date: 2016-07-29

* Fix SELECT JSON type cause UnicodeError
* Avoid float conversion while parsing microseconds
* Warning has number
* SSCursor supports warnings

## 0.7.5

Release date: 2016-06-28

* Fix exception raised while importing when getpwuid() fails (#472)
* SSCursor supports LOAD DATA LOCAL INFILE (#473)
* Fix encoding error happen for JSON type (#477)
* Fix test fail on Python 2.7 and MySQL 5.7 (#478)

## 0.7.4

Release date: 2016-05-26

* Fix AttributeError may happen while Connection.__del__ (#463)

pycloudwave/charset.py
MBLENGTH = {8: 1, 33: 3, 88: 2, 91: 2}


class Charset:
    def __init__(self, id, name, collation, is_default):
        self.id, self.name, self.collation = id, name, collation
        self.is_default = is_default == "Yes"

    def __repr__(self):
        return "Charset(id=%s, name=%r, collation=%r)" % (
            self.id,
            self.name,
            self.collation,
        )

    @property
    def encoding(self):
        name = self.name
        if name in ("utf8mb4", "utf8mb3"):
            return "utf8"
        if name == "latin1":
            return "cp1252"
        if name == "koi8r":
            return "koi8_r"
        if name == "koi8u":
            return "koi8_u"
        return name

    @property
    def is_binary(self):
        return self.id == 63


class Charsets:
    def __init__(self):
        self._by_id = {}
        self._by_name = {}

    def add(self, c):
        self._by_id[c.id] = c
        if c.is_default:
            self._by_name[c.name] = c

    def by_id(self, id):
        return self._by_id[id]

    def by_name(self, name):
        return self._by_name.get(name.lower())


_charsets = Charsets()
"""
Generated with:

cloudwave -N -s -e "select id, character_set_name, collation_name, is_default
from information_schema.collations order by id;" | python -c "import sys
for l in sys.stdin.readlines():
        id, name, collation, is_default  = l.split(chr(9))
        print '_charsets.add(Charset(%s, \'%s\', \'%s\', \'%s\'))' \
                % (id, name, collation, is_default.strip())
"

"""
_charsets.add(Charset(1, "big5", "big5_chinese_ci", "Yes"))
_charsets.add(Charset(2, "latin2", "latin2_czech_cs", ""))
_charsets.add(Charset(3, "dec8", "dec8_swedish_ci", "Yes"))
_charsets.add(Charset(4, "cp850", "cp850_general_ci", "Yes"))
_charsets.add(Charset(5, "latin1", "latin1_german1_ci", ""))
_charsets.add(Charset(6, "hp8", "hp8_english_ci", "Yes"))
_charsets.add(Charset(7, "koi8r", "koi8r_general_ci", "Yes"))
_charsets.add(Charset(8, "latin1", "latin1_swedish_ci", "Yes"))
_charsets.add(Charset(9, "latin2", "latin2_general_ci", "Yes"))
_charsets.add(Charset(10, "swe7", "swe7_swedish_ci", "Yes"))
_charsets.add(Charset(11, "ascii", "ascii_general_ci", "Yes"))
_charsets.add(Charset(12, "ujis", "ujis_japanese_ci", "Yes"))
_charsets.add(Charset(13, "sjis", "sjis_japanese_ci", "Yes"))
_charsets.add(Charset(14, "cp1251", "cp1251_bulgarian_ci", ""))
_charsets.add(Charset(15, "latin1", "latin1_danish_ci", ""))
_charsets.add(Charset(16, "hebrew", "hebrew_general_ci", "Yes"))
_charsets.add(Charset(18, "tis620", "tis620_thai_ci", "Yes"))
_charsets.add(Charset(19, "euckr", "euckr_korean_ci", "Yes"))
_charsets.add(Charset(20, "latin7", "latin7_estonian_cs", ""))
_charsets.add(Charset(21, "latin2", "latin2_hungarian_ci", ""))
_charsets.add(Charset(22, "koi8u", "koi8u_general_ci", "Yes"))
_charsets.add(Charset(23, "cp1251", "cp1251_ukrainian_ci", ""))
_charsets.add(Charset(24, "gb2312", "gb2312_chinese_ci", "Yes"))
_charsets.add(Charset(25, "greek", "greek_general_ci", "Yes"))
_charsets.add(Charset(26, "cp1250", "cp1250_general_ci", "Yes"))
_charsets.add(Charset(27, "latin2", "latin2_croatian_ci", ""))
_charsets.add(Charset(28, "gbk", "gbk_chinese_ci", "Yes"))
_charsets.add(Charset(29, "cp1257", "cp1257_lithuanian_ci", ""))
_charsets.add(Charset(30, "latin5", "latin5_turkish_ci", "Yes"))
_charsets.add(Charset(31, "latin1", "latin1_german2_ci", ""))
_charsets.add(Charset(32, "armscii8", "armscii8_general_ci", "Yes"))
_charsets.add(Charset(33, "utf8", "utf8_general_ci", "Yes"))
_charsets.add(Charset(34, "cp1250", "cp1250_czech_cs", ""))
_charsets.add(Charset(36, "cp866", "cp866_general_ci", "Yes"))
_charsets.add(Charset(37, "keybcs2", "keybcs2_general_ci", "Yes"))
_charsets.add(Charset(38, "macce", "macce_general_ci", "Yes"))
_charsets.add(Charset(39, "macroman", "macroman_general_ci", "Yes"))
_charsets.add(Charset(40, "cp852", "cp852_general_ci", "Yes"))
_charsets.add(Charset(41, "latin7", "latin7_general_ci", "Yes"))
_charsets.add(Charset(42, "latin7", "latin7_general_cs", ""))
_charsets.add(Charset(43, "macce", "macce_bin", ""))
_charsets.add(Charset(44, "cp1250", "cp1250_croatian_ci", ""))
_charsets.add(Charset(45, "utf8mb4", "utf8mb4_general_ci", "Yes"))
_charsets.add(Charset(46, "utf8mb4", "utf8mb4_bin", ""))
_charsets.add(Charset(47, "latin1", "latin1_bin", ""))
_charsets.add(Charset(48, "latin1", "latin1_general_ci", ""))
_charsets.add(Charset(49, "latin1", "latin1_general_cs", ""))
_charsets.add(Charset(50, "cp1251", "cp1251_bin", ""))
_charsets.add(Charset(51, "cp1251", "cp1251_general_ci", "Yes"))
_charsets.add(Charset(52, "cp1251", "cp1251_general_cs", ""))
_charsets.add(Charset(53, "macroman", "macroman_bin", ""))
_charsets.add(Charset(57, "cp1256", "cp1256_general_ci", "Yes"))
_charsets.add(Charset(58, "cp1257", "cp1257_bin", ""))
_charsets.add(Charset(59, "cp1257", "cp1257_general_ci", "Yes"))
_charsets.add(Charset(63, "binary", "binary", "Yes"))
_charsets.add(Charset(64, "armscii8", "armscii8_bin", ""))
_charsets.add(Charset(65, "ascii", "ascii_bin", ""))
_charsets.add(Charset(66, "cp1250", "cp1250_bin", ""))
_charsets.add(Charset(67, "cp1256", "cp1256_bin", ""))
_charsets.add(Charset(68, "cp866", "cp866_bin", ""))
_charsets.add(Charset(69, "dec8", "dec8_bin", ""))
_charsets.add(Charset(70, "greek", "greek_bin", ""))
_charsets.add(Charset(71, "hebrew", "hebrew_bin", ""))
_charsets.add(Charset(72, "hp8", "hp8_bin", ""))
_charsets.add(Charset(73, "keybcs2", "keybcs2_bin", ""))
_charsets.add(Charset(74, "koi8r", "koi8r_bin", ""))
_charsets.add(Charset(75, "koi8u", "koi8u_bin", ""))
_charsets.add(Charset(76, "utf8", "utf8_tolower_ci", ""))
_charsets.add(Charset(77, "latin2", "latin2_bin", ""))
_charsets.add(Charset(78, "latin5", "latin5_bin", ""))
_charsets.add(Charset(79, "latin7", "latin7_bin", ""))
_charsets.add(Charset(80, "cp850", "cp850_bin", ""))
_charsets.add(Charset(81, "cp852", "cp852_bin", ""))
_charsets.add(Charset(82, "swe7", "swe7_bin", ""))
_charsets.add(Charset(83, "utf8", "utf8_bin", ""))
_charsets.add(Charset(84, "big5", "big5_bin", ""))
_charsets.add(Charset(85, "euckr", "euckr_bin", ""))
_charsets.add(Charset(86, "gb2312", "gb2312_bin", ""))
_charsets.add(Charset(87, "gbk", "gbk_bin", ""))
_charsets.add(Charset(88, "sjis", "sjis_bin", ""))
_charsets.add(Charset(89, "tis620", "tis620_bin", ""))
_charsets.add(Charset(91, "ujis", "ujis_bin", ""))
_charsets.add(Charset(92, "geostd8", "geostd8_general_ci", "Yes"))
_charsets.add(Charset(93, "geostd8", "geostd8_bin", ""))
_charsets.add(Charset(94, "latin1", "latin1_spanish_ci", ""))
_charsets.add(Charset(95, "cp932", "cp932_japanese_ci", "Yes"))
_charsets.add(Charset(96, "cp932", "cp932_bin", ""))
_charsets.add(Charset(97, "eucjpms", "eucjpms_japanese_ci", "Yes"))
_charsets.add(Charset(98, "eucjpms", "eucjpms_bin", ""))
_charsets.add(Charset(99, "cp1250", "cp1250_polish_ci", ""))
_charsets.add(Charset(192, "utf8", "utf8_unicode_ci", ""))
_charsets.add(Charset(193, "utf8", "utf8_icelandic_ci", ""))
_charsets.add(Charset(194, "utf8", "utf8_latvian_ci", ""))
_charsets.add(Charset(195, "utf8", "utf8_romanian_ci", ""))
_charsets.add(Charset(196, "utf8", "utf8_slovenian_ci", ""))
_charsets.add(Charset(197, "utf8", "utf8_polish_ci", ""))
_charsets.add(Charset(198, "utf8", "utf8_estonian_ci", ""))
_charsets.add(Charset(199, "utf8", "utf8_spanish_ci", ""))
_charsets.add(Charset(200, "utf8", "utf8_swedish_ci", ""))
_charsets.add(Charset(201, "utf8", "utf8_turkish_ci", ""))
_charsets.add(Charset(202, "utf8", "utf8_czech_ci", ""))
_charsets.add(Charset(203, "utf8", "utf8_danish_ci", ""))
_charsets.add(Charset(204, "utf8", "utf8_lithuanian_ci", ""))
_charsets.add(Charset(205, "utf8", "utf8_slovak_ci", ""))
_charsets.add(Charset(206, "utf8", "utf8_spanish2_ci", ""))
_charsets.add(Charset(207, "utf8", "utf8_roman_ci", ""))
_charsets.add(Charset(208, "utf8", "utf8_persian_ci", ""))
_charsets.add(Charset(209, "utf8", "utf8_esperanto_ci", ""))
_charsets.add(Charset(210, "utf8", "utf8_hungarian_ci", ""))
_charsets.add(Charset(211, "utf8", "utf8_sinhala_ci", ""))
_charsets.add(Charset(212, "utf8", "utf8_german2_ci", ""))
_charsets.add(Charset(213, "utf8", "utf8_croatian_ci", ""))
_charsets.add(Charset(214, "utf8", "utf8_unicode_520_ci", ""))
_charsets.add(Charset(215, "utf8", "utf8_vietnamese_ci", ""))
_charsets.add(Charset(223, "utf8", "utf8_general_cloudwave500_ci", ""))
_charsets.add(Charset(224, "utf8mb4", "utf8mb4_unicode_ci", ""))
_charsets.add(Charset(225, "utf8mb4", "utf8mb4_icelandic_ci", ""))
_charsets.add(Charset(226, "utf8mb4", "utf8mb4_latvian_ci", ""))
_charsets.add(Charset(227, "utf8mb4", "utf8mb4_romanian_ci", ""))
_charsets.add(Charset(228, "utf8mb4", "utf8mb4_slovenian_ci", ""))
_charsets.add(Charset(229, "utf8mb4", "utf8mb4_polish_ci", ""))
_charsets.add(Charset(230, "utf8mb4", "utf8mb4_estonian_ci", ""))
_charsets.add(Charset(231, "utf8mb4", "utf8mb4_spanish_ci", ""))
_charsets.add(Charset(232, "utf8mb4", "utf8mb4_swedish_ci", ""))
_charsets.add(Charset(233, "utf8mb4", "utf8mb4_turkish_ci", ""))
_charsets.add(Charset(234, "utf8mb4", "utf8mb4_czech_ci", ""))
_charsets.add(Charset(235, "utf8mb4", "utf8mb4_danish_ci", ""))
_charsets.add(Charset(236, "utf8mb4", "utf8mb4_lithuanian_ci", ""))
_charsets.add(Charset(237, "utf8mb4", "utf8mb4_slovak_ci", ""))
_charsets.add(Charset(238, "utf8mb4", "utf8mb4_spanish2_ci", ""))
_charsets.add(Charset(239, "utf8mb4", "utf8mb4_roman_ci", ""))
_charsets.add(Charset(240, "utf8mb4", "utf8mb4_persian_ci", ""))
_charsets.add(Charset(241, "utf8mb4", "utf8mb4_esperanto_ci", ""))
_charsets.add(Charset(242, "utf8mb4", "utf8mb4_hungarian_ci", ""))
_charsets.add(Charset(243, "utf8mb4", "utf8mb4_sinhala_ci", ""))
_charsets.add(Charset(244, "utf8mb4", "utf8mb4_german2_ci", ""))
_charsets.add(Charset(245, "utf8mb4", "utf8mb4_croatian_ci", ""))
_charsets.add(Charset(246, "utf8mb4", "utf8mb4_unicode_520_ci", ""))
_charsets.add(Charset(247, "utf8mb4", "utf8mb4_vietnamese_ci", ""))
_charsets.add(Charset(248, "gb18030", "gb18030_chinese_ci", "Yes"))
_charsets.add(Charset(249, "gb18030", "gb18030_bin", ""))
_charsets.add(Charset(250, "gb18030", "gb18030_unicode_520_ci", ""))
_charsets.add(Charset(255, "utf8mb4", "utf8mb4_0900_ai_ci", ""))

charset_by_name = _charsets.by_name
charset_by_id = _charsets.by_id
